# **Shark Attack Data Analysis**

# Exploring Global Shark–Human Interaction Patterns

## Objectives

Project objectives:
- Analyse global shark attack incident data.
- Identify patterns in shark–human interactions.
- Explore geographic and behavioural factors associated with shark attacks.
- Develop an interactive dashboard communicating insights clearly.
- Evaluate ethical considerations related to wildlife incident data.
- Document the full data analytics project lifecycle.

Notebook objectives:
- Load and inspect the global shark attacks dataset.
- Clean and anonymise the dataset by removing personally identifiable information.
- Explore patterns in shark incidents across time, geography, and human activities.
- Analyse the relationship between activity type and fatality outcomes.
- Generate insights that will be used to build an interactive Power BI dashboard.


## Inputs

- data/raw/shark_attacks.csv – raw shark attack dataset obtained from Kaggle.
- Python libraries including pandas, numpy, matplotlib, seaborn, and plotly.
- Jupyter Notebook environment for data analysis and visualisation.


## Outputs

- Cleaned dataset for analysis.
- Exploratory data visualisations.
- Summary statistics describing shark attack patterns.
- Processed dataset exported for use in the Power BI dashboard.
- Visual outputs saved in outputs/figures.


## Additional Comments

- The dataset contains historical records of shark attack incidents from multiple countries.
- Personal identifiers such as victim names will be removed to ensure ethical data handling.
- The analysis focuses on identifying patterns in time, location, activity type, and fatality outcomes.
- Findings from this notebook will support the development of an interactive dashboard.



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\Shazia Mujahid\\Documents\\shark-attack-project\\shark-attack-data-dashboard\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\Shazia Mujahid\\Documents\\shark-attack-project\\shark-attack-data-dashboard'

# 1. Dataset Overview and Loading

In this section, the shark attack dataset is loaded and inspected to understand the structure. Initial checks are performed to examine the number of observations, available variables, missing values, and general dataset characteristics before moving onto the data cleaning stage. This step is crucial for identifying potential data quality issues before the data cleaning stage.


In [4]:
# Import libraries for data analysis and visualisation 

import pandas as pd     # Used for data manipulation and analysis
import numpy as np      # Provides numerical operations and array support
import matplotlib.pyplot as plt   # Used for creating basic visualisations
import seaborn as sns   # Built on matplotlib for improved statistical visualisations

In [5]:
# Load the raw shark attack dataset

df = pd.read_csv("data/raw/shark_attacks.csv")

# Display the first few rows to preview the dataset structure

df.head()

,date,year,type,country,area,location,activity,name,sex,age,fatal_y_n,time,species
0,2023-05-13,2023.0,Unprovoked,AUSTRALIA,South Australia,Elliston,Surfing,Simon Baccanello,M,46,Y,10h10,White shark
1,2023-04-29,2023.0,Unprovoked,AUSTRALIA,Western Australia,"Yallingup, Busselton",Swimming,male,M,NaN,N,11h20,1m shark
2,2022-10-07,2022.0,Unprovoked,AUSTRALIA,Western Australia,Port Hedland,Spearfishing,Robbie Peck,M,38,N,11h30,Bull shark
3,2021-10-04,2021.0,Unprovoked,USA,Florida,"Fort Pierce State Park, St. Lucie County",Surfing,Truman Van Patrick,M,25.0,N,NaN,NaN
4,2021-10-03,2021.0,Unprovoked,USA,Florida,"Jensen Beach, Martin County",Swimming,male,M,NaN,N,12h00,NaN


In [6]:
# Examine the size of the dataset
# This shows the total number of rows (records) and columns (variables)

df.shape

(6890, 13)

The dataset contains 6,890 recorded shark incident observations and 13 variables describing each incident. Each row represents a single shark attack record, while the columns contain information such as the date, location, activity being performed, shark species involved, and whether the incident resulted in a fatality.

In [7]:
# Display the column names to understand which variables are available
# This helps identify which variables may be relevant for analysis

df.columns

Index(['date', 'year', 'type', 'country', 'area', 'location', 'activity',
       'name', 'sex', 'age', 'fatal_y_n', 'time', 'species'],
      dtype='object')

The dataset contains 13 variables describing each shark incident. These variables include temporal information such as the date and year of the incident, geographic information including country and location, and contextual details such as the activity being performed and the shark species involved.

The dataset also contains demographic attributes of individuals involved in the incidents and an indicator of whether the incident resulted in a fatality.

In [8]:
# Display dataset information including data types and non-null values
# This helps identify missing data and potential type conversions required during cleaning

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6890 entries, 0 to 6889
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   date       6587 non-null   object 
 1   year       6758 non-null   float64
 2   type       6871 non-null   object 
 3   country    6839 non-null   object 
 4   area       6409 non-null   object 
 5   location   6325 non-null   object 
 6   activity   6304 non-null   object 
 7   name       6670 non-null   object 
 8   sex        6318 non-null   object 
 9   age        3903 non-null   object 
 10  fatal_y_n  6890 non-null   object 
 11  time       3372 non-null   object 
 12  species    3772 non-null   object 
dtypes: float64(1), object(12)
memory usage: 699.9+ KB


The dataset contains 6,890 observations and 13 variables describing shark incident records. Most variables are stored as object (string) data types, while the year variable is stored as a numeric value (float64).

Several variables contain missing values, which will need to be addressed during the data cleaning stage. For example, the variables age, time, and species contain a large number of missing entries.

The variable name contains personally identifiable information and will be removed during the cleaning process to ensure the dataset is anonymised before analysis.

In [9]:
# Check the number of missing values in each column
# Missing data may require cleaning or transformation before analysis

df.isnull().sum()

date          303
year          132
type           19
country        51
area          481
location      565
activity      586
name          220
sex           572
age          2987
fatal_y_n       0
time         3518
species      3118
dtype: int64

The missing value analysis shows that several variables in the dataset contain incomplete information. Some variables such as type, country, and year contain relatively small numbers of missing values and can be retained with minimal cleaning.

However, other variables contain substantial missing data. For example, the age, time, and species columns contain a large proportion of missing values. These variables may require careful handling or may be excluded from certain analyses depending on their completeness.

As previously mentioned, the name column contains personally identifiable information and will be removed during the cleaning stage to ensure the dataset used for analysis is anonymised.

---

# 2. Data Preparation

This section prepares the dataset for further analysis. Initial preparation steps are performed to ensure the dataset is suitable for cleaning and analysis.

These steps include removing variables that are not required for the analysis, preparing column names for easier manipulation, and converting certain variables into more appropriate data types. Preparing the dataset in this way helps ensure consistency and improves the efficiency of subsequent data cleaning and feature engineering steps.

## 2.1 Remove Personally Identifiable Information

The dataset contains a column labelled `name`, which includes the names of individuals involved in shark attack incidents. Since this information is not required for the analysis and represents personally identifiable information, the column is removed to maintain data privacy and ensure the dataset used for analysis is anonymised.

In [10]:
# Remove the 'name' column from the dataset.
# This column contains personally identifiable information (PII), which is not needed for the analysis and should be excluded to help preserve privacy and anonymise the dataset.

df = df.drop(columns=["name"])

# Display the first few rows to confirm that the column has been removed from the DataFrame.

df.head()

,date,year,type,country,area,location,activity,sex,age,fatal_y_n,time,species
0,2023-05-13,2023.0,Unprovoked,AUSTRALIA,South Australia,Elliston,Surfing,M,46,Y,10h10,White shark
1,2023-04-29,2023.0,Unprovoked,AUSTRALIA,Western Australia,"Yallingup, Busselton",Swimming,M,NaN,N,11h20,1m shark
2,2022-10-07,2022.0,Unprovoked,AUSTRALIA,Western Australia,Port Hedland,Spearfishing,M,38,N,11h30,Bull shark
3,2021-10-04,2021.0,Unprovoked,USA,Florida,"Fort Pierce State Park, St. Lucie County",Surfing,M,25.0,N,NaN,NaN
4,2021-10-03,2021.0,Unprovoked,USA,Florida,"Jensen Beach, Martin County",Swimming,M,NaN,N,12h00,NaN


## 2.2 Standardise Column Names

To improve readability and ensure consistency when referencing variables in code, column names are standardised by converting them to lowercase and removing any unnecessary whitespace.

In [11]:
# Standardise all column names by:
# 1. converting them to lowercase
# 2. removing leading and trailing whitespace

# This makes the dataset easier to work with and reduces the risk of coding errors caused by inconsistent naming conventions.

df.columns = df.columns.str.lower().str.strip()

# Display the updated column names to verify that the formatting has been applied correctly.
df.columns

Index(['date', 'year', 'type', 'country', 'area', 'location', 'activity',
       'sex', 'age', 'fatal_y_n', 'time', 'species'],
      dtype='object')

## 2.3 Convert the Year Variable

The `year` variable is currently stored as a floating-point data type. Since years are whole-number values, this variable is converted to numeric format to ensure consistency and support time-based analysis.

In [12]:
# Convert the 'year' column to a numeric data type.
# Any non-numeric values will be converted to NaN using errors='coerce'.

df["year"] = pd.to_numeric(df["year"], errors="coerce")

# Convert the column to pandas nullable integer type (Int64).
# This allows integer values while still supporting missing values.

df["year"] = df["year"].astype("Int64")

# Check the data type to confirm the conversion.

df["year"].dtype

Int64Dtype()

## 2.4 Convert the Date Variable

The `date` column is converted from text format to a datetime format. This ensures the variable is suitable for chronological analysis and allows dates to be handled more efficiently in pandas.

In [13]:
# Convert the 'date' column to datetime format.
# Invalid or inconsistent date values will be converted to NaT (Not a Time) using errors='coerce'.

# This allows the date variable to be used for time-based analysis, filtering, and sorting.

df["date"] = pd.to_datetime(df["date"], errors="coerce")

# Display the first few values from the 'date' column to verify that the conversion was successful.

df["date"].head()

0   2023-05-13
1   2023-04-29
2   2022-10-07
3   2021-10-04
4   2021-10-03
Name: date, dtype: datetime64[ns]

## 2.5 Verify Dataset Structure

After the initial preparation steps, the dataset structure is reviewed to confirm that the transformations have been applied correctly and that the dataset is ready for the cleaning stage.

In [14]:
# Display a summary of the dataset structure after preparation.
# This includes:
# - the number of rows and columns
# - column names
# - non-null counts
# - data types

# This step is used to confirm that:
# - the 'name' column has been removed
# - the 'date' column is now datetime
# - the 'year' column has been converted to numeric format

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6890 entries, 0 to 6889
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   date       6577 non-null   datetime64[ns]
 1   year       6758 non-null   Int64         
 2   type       6871 non-null   object        
 3   country    6839 non-null   object        
 4   area       6409 non-null   object        
 5   location   6325 non-null   object        
 6   activity   6304 non-null   object        
 7   sex        6318 non-null   object        
 8   age        3903 non-null   object        
 9   fatal_y_n  6890 non-null   object        
 10  time       3372 non-null   object        
 11  species    3772 non-null   object        
dtypes: Int64(1), datetime64[ns](1), object(10)
memory usage: 652.8+ KB


### Section 2 Summary

The dataset has now undergone initial preparation steps. Personally identifiable information has been removed, and key variables such as `date` and `year` have been formatted to support analysis.

These preparation steps improve dataset consistency and ensure that the data is ready for the next stage, which focuses on cleaning missing values, resolving inconsistencies, and engineering features for analysis.

---

# 3. Data Cleaning and Feature Engineering

This section focuses on cleaning the dataset and preparing variables for analysis. Real-world datasets often contain inconsistencies, missing values, and formatting issues that must be addressed before meaningful analysis can be conducted.

The cleaning process includes handling missing values, standardising categorical variables, and improving the overall quality of the dataset. These steps ensure that the dataset is reliable and suitable for exploratory analysis and visualisation in later stages of the project.

## 3.1 Review Missing Values

Before performing cleaning operations, the dataset is re-examined to identify the extent of missing data across each variable. Understanding where missing values occur helps guide decisions on whether data should be removed, imputed, or retained depending on analytical relevance.

In [15]:
# Recalculate the number of missing values in each column.
# This provides an updated overview of dataset completeness before applying cleaning operations.

df.isnull().sum()

date          313
year          132
type           19
country        51
area          481
location      565
activity      586
sex           572
age          2987
fatal_y_n       0
time         3518
species      3118
dtype: int64

The missing value analysis highlights several variables with incomplete data. Some variables such as `type`, `country`, and `year` contain relatively few missing values and can be retained without significant impact on the analysis.

Other variables, including `activity`, `sex`, `area`, and `location`, contain moderate levels of missing data but still provide useful contextual information about shark attack incidents.

However, variables such as `age`, `time`, and `species` contain a substantial number of missing values. While these variables remain potentially valuable for analysis, missing entries will need to be considered when interpreting results.

Also, the `fatal_y_n` variable contains no missing values, making it a reliable variable for analysing the outcomes of shark attack incidents.

## 3.2 Remove Records Without a Year

The `year` variable is critical for analysing temporal trends in shark attacks. Records that do not contain a valid year value cannot contribute to time-based analysis and are therefore removed from the dataset.

In [16]:
# Remove rows where the year value is missing.
# These records cannot be used for time-based analysis.

df = df.dropna(subset=["year"])

# Confirm the dataset size after removal.

df.shape

(6758, 12)

In [ ]:
# Validate that there are no missing values in the 'year' column after cleaning.

df["year"].isnull().sum()

0

Records with missing values in the `year` column were removed because the year variable is essential for analysing temporal trends in shark attack incidents.

A total of 132 records were removed from the dataset, reducing the dataset from 6,890 to 6,758 observations. Removing these records ensures that all remaining entries contain valid temporal information required for time-based analysis.

## 3.3 Standardise the Sex Variable

The `sex` variable contains inconsistent representations of gender, such as "M", "Male", or lowercase variations. These values are standardised to ensure consistent categorisation during analysis.

In [18]:
# View the unique values in the sex column to understand existing variations.

df["sex"].unique()

array(['M', 'F', nan, 'N', '.', 'M x 2', 'lli'], dtype=object)

In [19]:
# Standardise the sex column by converting values to uppercase to ensure consistent formatting.

df["sex"] = df["sex"].str.upper()

# Replace invalid or ambiguous entries with NaN so that only valid categories remain.

df["sex"] = df["sex"].replace(["N", ".", "M X 2", "LLI"], np.nan)

# Check the cleaned values

df["sex"].unique()

array(['M', 'F', nan], dtype=object)

The `sex` variable contained several inconsistent or invalid entries, including values such as "N", ".", and other irregular text strings. These values likely represent data entry errors or ambiguous records.

To ensure consistency, the variable was standardised by converting text to uppercase and replacing invalid entries with missing values. After cleaning, the column contains only valid categories representing male (`M`) and female (`F`) individuals, with missing values where the information was unavailable.

## 3.4 Clean the Age Variable

The `age` column is currently stored as a text variable and may contain non-numeric values. To support quantitative analysis, the column is converted into a numeric format. Any invalid values are converted to missing values.

In [20]:
# Convert the age column to numeric format.
# Invalid entries will be converted to NaN using errors='coerce'.

df["age"] = pd.to_numeric(df["age"], errors="coerce")

# Verify the updated data type.

df["age"].dtype

dtype('float64')

The `age` column was originally stored as a text variable and contained non-numeric values. To support quantitative analysis, the column was converted to numeric format using `pd.to_numeric()`.

Any invalid or non-numeric values were automatically converted to missing values (`NaN`). After conversion, the variable can now be used for statistical analysis and visualisation of age-related patterns in shark attack incidents.

## 3.5 Inspect Activity Categories

The `activity` variable describes the human activity being performed at the time of the incident. Since this field contains free-text descriptions, it may include a wide variety of values that require standardisation.

In [21]:
# View the most common activity values to understand the types of activities recorded.

df["activity"].value_counts().head(20)

activity
Surfing            1111
Swimming            993
Fishing             490
Spearfishing        381
Wading              169
Bathing             163
Diving              133
Snorkeling          126
Standing            113
Scuba diving         82
Body boarding        64
Body surfing         51
Boogie boarding      42
Kayaking             41
Fell overboard       33
Treading water       33
Free diving          30
Pearl diving         28
Surf skiing          24
Windsurfing          20
Name: count, dtype: int64

In [22]:
# Create a simplified activity category variable. This groups similar activities together to make analysis easier.

df["activity_group"] = df["activity"].str.lower()

# Surfing related activities

df.loc[df["activity_group"].str.contains("surf", na=False), "activity_group"] = "surfing"


# Swimming related activities

df.loc[df["activity_group"].str.contains("swim|bathing|wading|standing|treading", na=False), "activity_group"] = "swimming"


# Fishing related activities

df.loc[df["activity_group"].str.contains("fish", na=False), "activity_group"] = "fishing"


# Diving related activities

df.loc[df["activity_group"].str.contains("div", na=False), "activity_group"] = "diving"


# Watercraft activities

df.loc[df["activity_group"].str.contains("kayak|boat|ski|wind|board", na=False), "activity_group"] = "watercraft"


# Everything else

df.loc[~df["activity_group"].isin(["surfing","swimming","fishing","diving","watercraft"]), "activity_group"] = "other"


# Check the new grouped categories

df["activity_group"].value_counts()

activity_group
swimming      1745
surfing       1417
other         1404
fishing       1243
diving         514
watercraft     435
Name: count, dtype: int64

In [23]:
# Validate that the new activity group variable has been created correctly by displaying the original and new columns side by side for the first few records.

df[["activity", "activity_group"]].head(10)

,activity,activity_group
0,Surfing,surfing
1,Swimming,swimming
2,Spearfishing,fishing
3,Surfing,surfing
4,Swimming,swimming
5,Swimming,swimming
6,Surfing,surfing
7,Spearfishing,fishing
8,NaN,other
9,Swimming,swimming


The `activity` variable contained a large number of unique free-text descriptions describing human activities at the time of the incident. Many of these entries referred to similar activities but were recorded using different wording.

To simplify the analysis and improve interpretability, activities were grouped into broader categories including swimming, surfing, fishing, diving, and watercraft-related activities. Any remaining activities that did not fit into these categories were classified as "other".

This grouping reduces the number of activity categories and allows clearer comparisons when analysing patterns in shark attack incidents.

## 3.6 Inspect Shark Species

The `species` variable identifies the shark species involved in each incident. However, this field contains many variations and descriptive entries. Examining the most common values helps determine whether grouping or simplification may be required.

In [24]:
# Display the most common shark species entries to identify inconsistencies or variations.

df["species"].value_counts().head(20)

species
White shark                                           191
Shark involvement prior to death was not confirmed    105
Invalid                                               101
Shark involvement not confirmed                        92
Tiger shark                                            86
Shark involvement prior to death unconfirmed           68
Bull shark                                             62
4' shark                                               43
6' shark                                               42
1.8 m [6'] shark                                       35
Questionable incident                                  35
Questionable                                           33
1.5 m [5'] shark                                       32
5' shark                                               29
Wobbegong shark                                        28
4' to 5' shark                                         27
3' shark                                               27
1.2 m 

In [25]:
# Create a lowercase version of the species column to make text matching easier.

df["species_group"] = df["species"].str.lower()


# Identify major species

df.loc[df["species_group"].str.contains("white", na=False), "species_group"] = "white shark"
df.loc[df["species_group"].str.contains("tiger", na=False), "species_group"] = "tiger shark"
df.loc[df["species_group"].str.contains("bull", na=False), "species_group"] = "bull shark"


# All other entries containing the word "shark" are grouped as "other shark"

df.loc[df["species_group"].str.contains("shark", na=False), "species_group"] = "other shark"


# Assign the major species again (so they are not overwritten)

df.loc[df["species"].str.contains("White", na=False), "species_group"] = "white shark"
df.loc[df["species"].str.contains("Tiger", na=False), "species_group"] = "tiger shark"
df.loc[df["species"].str.contains("Bull", na=False), "species_group"] = "bull shark"


# Replace missing or uncertain values

df.loc[df["species_group"].isna(), "species_group"] = "unknown"


# Check the grouped results

df["species_group"].value_counts()

species_group
unknown                                                                                                                                                          3024
other shark                                                                                                                                                      2589
white shark                                                                                                                                                       491
tiger shark                                                                                                                                                       263
bull shark                                                                                                                                                        157
invalid                                                                                                                                                     

In [26]:
# Replace invalid or uncertain species labels with "unknown"

df["species_group"] = df["species_group"].replace([
    "invalid",
    "questionable incident",
    "questionable",
    "unidentified species",
    "unidentified"
], "unknown")


# Check the cleaned results again

df["species_group"].value_counts()

species_group
unknown                                                                                                                                                          3207
other shark                                                                                                                                                      2589
white shark                                                                                                                                                       491
tiger shark                                                                                                                                                       263
bull shark                                                                                                                                                        157
bronze whaler                                                                                                                                               

In [27]:
# Define the allowed species categories

major_species = ["white shark", "tiger shark", "bull shark", "other shark", "unknown"]


# Any value not in the major species list will be grouped as "other shark"

df.loc[~df["species_group"].isin(major_species), "species_group"] = "other shark"


# Check the final result

df["species_group"].value_counts()

species_group
unknown        3207
other shark    2640
white shark     491
tiger shark     263
bull shark      157
Name: count, dtype: int64

The `species` column contained numerous inconsistent entries including size descriptions, uncertain identifications, and free-text incident notes. To improve interpretability, the variable was simplified by grouping shark species into broader categories.

Major species such as white sharks, tiger sharks, and bull sharks were retained as separate categories due to their frequency in the dataset. All other shark-related entries were grouped under "other shark", while records without confirmed shark identification were classified as "unknown".

As a result, this approach reduces the complexity of the species variable and allows clearer comparisons in subsequent analysis and visualisations.

## 3.7 Verify Dataset After Cleaning

After applying initial cleaning steps, the dataset structure is reviewed to confirm that transformations have been applied successfully and that the dataset is ready for exploratory analysis.

In [28]:
# Review the structure of the cleaned dataset.
# The info() function displays the number of rows, columns, non-null counts, and data types for each variable.
# This step confirms that the previous cleaning operations (such as removing rows, converting data types, and creating new variables) were applied successfully.

df.info()


# Display the first few rows of the dataset to visually confirm that the cleaned variables (such as activity_group and species_group) have been created correctly and that the dataset appears consistent.

df.head()


# Recalculate the number of missing values in each column. This final check helps verify that missing values have been handled appropriately and highlights any remaining incomplete variables that may need to be considered during analysis.

df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
Index: 6758 entries, 0 to 6884
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   date            6575 non-null   datetime64[ns]
 1   year            6758 non-null   Int64         
 2   type            6742 non-null   object        
 3   country         6712 non-null   object        
 4   area            6305 non-null   object        
 5   location        6230 non-null   object        
 6   activity        6194 non-null   object        
 7   sex             6190 non-null   object        
 8   age             3763 non-null   float64       
 9   fatal_y_n       6758 non-null   object        
 10  time            3365 non-null   object        
 11  species         3735 non-null   object        
 12  activity_group  6758 non-null   object        
 13  species_group   6758 non-null   object        
dtypes: Int64(1), datetime64[ns](1), float64(1), object(11)
memory

date               183
year                 0
type                16
country             46
area               453
location           528
activity           564
sex                568
age               2995
fatal_y_n            0
time              3393
species           3023
activity_group       0
species_group        0
dtype: int64

The cleaned dataset was reviewed to verify that all transformations were applied correctly. The dataset now contains 6758 observations and 14 variables, including newly created variables for grouped activity types and shark species.

Key variables such as `year`, `activity_group`, and `species_group` are now complete and suitable for analysis. While some variables such as `age`, `time`, and `species` still contain missing values, these variables are not critical for all analyses and can still provide useful insights where data is available.

Overall, this validation step confirms that the dataset has been successfully cleaned and prepared for exploratory data analysis.

---